### PR TITLE
NIAD-1596: Immunization: only 1st reason given processed into notes text

### PR DESCRIPTION
**Pull request by bbell-kainos**

### CODE REVIEW

**Review by rorysproule1 — DISMISSED** [review]

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java [hunk 1]
```diff
@@ -217,9 +217,12 @@ private String buildNotePertinentInformation(Immunization immunization) {
...
         if (immunization.hasExplanation() && immunization.getExplanation().hasReason()) {
-            CodeableConcept reason = immunization.getExplanation().getReasonFirstRep();
-            explanation = CodeableConceptMappingUtils.extractTextOrCoding(reason);
-            return explanation.map(value -> REASON + value).orElse(StringUtils.EMPTY);
+            String reasonGiven = REASON + immunization.getExplanation().getReason().stream()
+                .map(CodeableConceptMappingUtils::extractTextOrCoding)
+                .map(Optional::get)
+                .filter(value -> !value.equals(StringUtils.EMPTY))
+                .collect(Collectors.joining(StringUtils.SPACE));
+            return reasonGiven.equals(REASON) ? StringUtils.EMPTY : reasonGiven;
```
Comment (bartek-sarul) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java:225]: This is a bit unsafe. I would build a string with reasons and add the REASON only if the list is not empty.
```
return StringUtils.EMPTY.equals(reasons) ? StringUtils.EMPTY : (REASON + reasons);
```

**Review by bartek-sarul — CHANGES_REQUESTED** [review]

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java (hunk 1 above)
Comment (bbell-kainos) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java]: Addressed

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java [hunk 2]
```diff
@@ -217,9 +217,12 @@ private String buildNotePertinentInformation(Immunization immunization) {
     private String buildExplanationPertinentInformation(Immunization immunization) {
         Optional<String> explanation;
         if (immunization.hasExplanation() && immunization.getExplanation().hasReason()) {
-            CodeableConcept reason = immunization.getExplanation().getReasonFirstRep();
-            explanation = CodeableConceptMappingUtils.extractTextOrCoding(reason);
-            return explanation.map(value -> REASON + value).orElse(StringUtils.EMPTY);
+            String reasonGiven = immunization.getExplanation().getReason().stream()
+                .map(CodeableConceptMappingUtils::extractTextOrCoding)
+                .map(Optional::get)
```
Comment (bartek-sarul) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java:222]: It's not safe to directly call get on an Optional. Better to use this
```
String reasonNotGiven = immunization.getExplanation().getReason().stream()
    .map(CodeableConceptMappingUtils::extractTextOrCoding)
    .flatMap(Optional::stream)
    .filter(StringUtils::isNotBlank)
    .collect(Collectors.joining(StringUtils.SPACE));
```
I've also included `.filter(StringUtils::isNotBlank)` in the code example

**Review by bartek-sarul — CHANGES_REQUESTED** [review]

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java (hunk 2 above)
Comment (bbell-kainos) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java]: Addressed and applied to related PR: https://github.com/nhsconnect/integration-adaptor-gp2gp/pull/320

**Review by rorysproule1 — APPROVED** [review]

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java [hunk 3]
```diff
@@ -217,9 +217,12 @@ private String buildNotePertinentInformation(Immunization immunization) {
...
         if (immunization.hasExplanation() && immunization.getExplanation().hasReason()) {
-            CodeableConcept reason = immunization.getExplanation().getReasonFirstRep();
-            explanation = CodeableConceptMappingUtils.extractTextOrCoding(reason);
-            return explanation.map(value -> REASON + value).orElse(StringUtils.EMPTY);
+            String reasonGiven = immunization.getExplanation().getReason().stream()
+                .map(CodeableConceptMappingUtils::extractTextOrCoding)
+                .flatMap(Optional::stream)
+                .filter(StringUtils::isNotBlank)
+                .collect(Collectors.joining(StringUtils.SPACE));
+            return StringUtils.EMPTY.equals(reasonGiven) ? StringUtils.EMPTY : (REASON + reasonGiven);
```
Comment (bartek-sarul) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java:225]: You should check this using `StringUtils.isBlank`

**Review by bartek-sarul — CHANGES_REQUESTED** [review]

**Review by bartek-sarul — APPROVED** [review]

**Review by ElliotB-96 — APPROVED** [review]

**Pull request merged**